### ISSUE TITLE
Load Execution returning EmptyLoad even if valid records available after filtering

### ISSUE THREAD

**Issue by SajeevAr** [CommBank/maestro#385]
In a file based load, the type returned in LoadInfo is used for handling different load scenarios. But LoadExecution is returning an EmptyLoad even though valid records are available after filtering. While running the same job again with same input files, it may return LoadSuccess. This issue is occurring only randomly.

It seems this is due to a TimeoutException occurring in HadoopStepStats, while attempting to fetch counter values for the job. TimeoutException is handled in HadoopStepStats.java and it only logs a warning as shown below. The warning is only found in logs where this issue has occured and it is not seen for any other valid EmptyLoad/LoadSuccess scenarios.

Please find the relevant log below :

15/05/06 14:30:58 INFO flow.Flow: [] starting
15/05/06 14:30:58 INFO flow.Flow: []  source: Hfs["TextLine[['offset', 'line']->[ALL]]"]["/dev/source/commsee/car/CSE_CCA_PRICING_OPTION/2015/05/05"]
15/05/06 14:30:58 INFO flow.Flow: []  sink: Hfs["CombinedSequenceFileScheme[[0:1]]"]["/tmp/hadoop-acoe_omnds_dintg/maestro/7e5cbe13-fdb1-47c7-8d96-eca60fad71a6.seq"]
15/05/06 14:30:58 INFO flow.Flow: []  parallel execution is enabled: true
15/05/06 14:30:58 INFO flow.Flow: []  starting jobs: 1
15/05/06 14:30:58 INFO flow.Flow: []  allocating threads: 1
15/05/06 14:30:58 INFO flow.FlowStep: [] starting step: (1/1) ...7c7-8d96-eca60fad71a6.seq
15/05/06 14:31:05 INFO mapred.FileInputFormat: Total input paths to process : 1
15/05/06 14:31:19 INFO flow.FlowStep: [] submitted hadoop job: job_201505030350_10719
15/05/06 14:31:19 INFO flow.FlowStep: [] tracking url: 
15/05/06 14:32:28 WARN hadoop.HadoopStepStats: fetching counters timed out after: 5 seconds, attempts: 1
15/05/06 14:32:28 INFO util.Hadoop18TapUtil: deleting temp path /tmp/hadoop-acoe_omnds_dintg/maestro/7e5cbe13-fdb1-47c7-8d96-eca60fad71a6.seq/_temporary
15/05/06 14:32:28 INFO car.CseCcaPricingOptionExecution$: Creating empty outbound partition /dev/etl/outbound/commsee/car/CSE_CCA_PRICING_OPTION/year=2015/month=05/day=05
15/05/06 14:32:28 INFO Configuration.deprecation: io.bytes.per.checksum is deprecated. Instead, use dfs.bytes-per-checksum
Empty File Load
Exited job au.com.cba.omnia.etl.commsee.car.CseCcaPricingOptionExecution (0.1.0-20150506042246-926bb79) at Wed May  6 14:32:33 AEST 2015 with status 0

**Comment by rowandavies** [comment]
@SajeevAr I think you said earlier that there was only one known occurrence of this situation.  Is this still the case?  Is it possible to repeatedly run the same load to see how often it occurs again?

I think you investigated this issue some, including the Cascading source code, and thought a bit about what we might do.  Can you add anything you think would be useful here just so no one has to repeat what you've already done?

**Comment by vineethvarghese** [comment]
A quick look at the cascading code suggests that we can control the wait time to get a counter. The default is 5 sec which might be too low.

**Comment by SajeevAr** [comment]
@rowandavies Unfortunately it is very difficult to recreate this scenario, since it is hard to predict when this timeout will occur. I tried to run the same job multiple times, but this issue occurred only once for the current load, but seen the same issue in a different project some time before. So this may require some deeper investigation.

Though this occurs rarely, if it is not handled, it can affect the downstream jobs since no data will be loaded into tables in such scenario.

When this issue occurred, even if LoadInfo returned EmptyLoad, the job tracker logs indicated correct values for cascading.flow.StepCounters - Tuples_Read, Tuples_Written and also for tuples_filtered.

**Comment by SajeevAr** [comment]
Another occurrence of the same issue happened for the same load with a different input set. This time also we can seen time out warning in the log.

15/05/21 14:44:45 INFO flow.FlowStep: [] starting step: (1/1) ...b8a-8330-eb6674466b84.seq
15/05/21 14:44:57 INFO mapred.FileInputFormat: Total input paths to process : 1
15/05/21 14:44:58 INFO Configuration.deprecation: io.bytes.per.checksum is deprecated. Instead, use dfs.bytes-per-checksum
15/05/21 14:45:02 INFO flow.FlowStep: [] submitted hadoop job: job_201505121835_22116
15/05/21 14:45:02 INFO flow.FlowStep: [] tracking url: 
15/05/21 14:45:50 WARN hadoop.HadoopStepStats: fetching counters timed out after: 5 seconds, attempts: 1
15/05/21 14:45:50 INFO util.Hadoop18TapUtil: deleting temp path /tmp/hadoop-acoe_omnds_dintg/maestro/3dc84add-18f1-4b8a-8330-eb6674466b84.seq/_temporary
15/05/21 14:45:50 ERROR car.CseCcaAppExecution$: Mismatch in trailer count and actual data loaded
Exited job au.com.cba.omnia.etl.commsee.car.CseCcaAppExecution (0.1.0-20150521044053-f98fcfc) at Thu May 21 14:46:24 AEST 2015 with status 255

**Comment by vineethvarghese** [comment]
@stephanh Maybe we should take this up?

**Comment by vineethvarghese** [comment]
@SajeevAr can you please paste the logs for `When this issue occurred, even if LoadInfo returned EmptyLoad, the job tracker logs indicated correct values for cascading.flow.StepCounters - Tuples_Read, Tuples_Written and also for tuples_filtered.` if you have them.

**Comment by SajeevAr** [comment]
@vineethvarghese I don't have the complete job tracker log with me now. But I have the following screenshots indicating the counter values for this job. If I encounter this issue again, I shall take the complete log.

![image](https://cloud.githubusercontent.com/assets/10343887/7852373/e53729b6-053f-11e5-9bc9-4632f0f87932.png)

![image](https://cloud.githubusercontent.com/assets/10343887/7852378/f4bd97f8-053f-11e5-8e97-e8253c8b50f4.png)

**Comment by vineethvarghese** [comment]
Thanks. I was more interested to see all the logs rather than actual counts
to decide on the timeout. We have a fix that increases the timeout to get
the counter.  Hopefully this will address your issue.

**Comment by stephanh** [comment]
We believe this problem is caused by the job tracker being to slow to respond to the request from the job to get the counter. We have raised an incident `IM1117666` to get this looked into and in the mean time we are going to increase the counter time out in the latest version of Maestro (see #402).

**Comment by vineethvarghese** [comment]
We interact with the hadoop platform via scalding and cascading. After a Maestro load or view, a count is returned by writing the file to a temporary location and then getting the hadoop job counters to find the number of records written. This seemingly one liner breaks down to a cascading flow and a flow step which in turn maps to MR job on the cluster. On completion of the job, we attempt to get the job counter via `scalding -> cascading -> hadoop job tracker`. The cascading api tries to do the reasonable thing and timeouts the request after the job tracker hasn't responded in 5 seconds. The fact that the job tracker is not responding in 5 seconds is a problem considering that the file has already been copied. 

As @stephanh mentioned, we have raised an incident for the platform team to investigate the job tracker issue and in the meantime bumped up the counter timeout to 60 sec.

**Comment by vineethvarghese** [comment]
This is fix is on master now, so closing this. Please try out the latest version of maestro and reopen this issue if you still see the problem.

**Comment by maxmithun** [comment]
Currently we are experiencing the issue of getting timeout getting counters with updated `util.etl version 1.12.1-20150528110707-248b50b` also.

```
15/07/08 18:32:07 INFO hive.metastore: Trying to connect to metastore with URI thrift://r01wn01.nw-omnia.cba:9083
15/07/08 18:32:07 INFO hive.metastore: Connected to metastore.
15/07/08 18:32:07 INFO Configuration.deprecation: io.bytes.per.checksum is deprecated. Instead, use dfs.bytes-per-checksum
15/07/08 18:32:07 INFO util.HadoopUtil: using default application jar, may cause class not found exceptions on the cluster
15/07/08 18:32:07 INFO planner.HadoopPlanner: using application jar: /disk1/bigdata/dev/scripts/farlo/1.0.0-20150624014544-cd4bd39/etl-farlo-clm-assembly-1.0.0-20150708030821-b6ae56e.jar
15/07/08 18:32:07 INFO flow.Flow: [] starting
15/07/08 18:32:07 INFO flow.Flow: []  source: Hfs["CombinedSequenceFileScheme[[0:1]]"]["/tmp/hadoop-acoe_omnds_dintg/maestro/c7e86177-a3a0-4363-8695-5cb4172d6861.seq"]
15/07/08 18:32:07 INFO flow.Flow: []  sink: PartitionHiveParquetWriteTap["HiveTap["ParquetScroogeScheme[[UNKNOWN]->['0']]"]["hdfs://nameservice1/dev/view/warehouse/farlo/clm/v_CLNT_GRUP_STGY_HIST"]"]["HivePartition{delimiter='/'partitionFields=''year', 'month''}"]
15/07/08 18:32:07 INFO flow.Flow: []  parallel execution is enabled: true
15/07/08 18:32:07 INFO flow.Flow: []  starting jobs: 1
15/07/08 18:32:07 INFO flow.Flow: []  allocating threads: 1
15/07/08 18:32:07 INFO flow.FlowStep: [] starting step: (1/1) ...v_CLNT_GRUP_STGY_HIST/*/*
15/07/08 18:32:19 INFO mapred.FileInputFormat: Total input paths to process : 3
15/07/08 18:32:32 INFO flow.FlowStep: [] submitted hadoop job: job_201506272351_50208
15/07/08 18:32:32 INFO flow.FlowStep: [] tracking url: http://r01hn02.nw-omnia.cba:50030/jobdetails.jsp?jobid=job_201506272351_50208
15/07/08 18:34:21 INFO hive.metastore: Trying to connect to metastore with URI thrift://r01wn01.nw-omnia.cba:9083
15/07/08 18:34:21 INFO hive.metastore: Connected to metastore.
15/07/08 18:34:21 INFO hive.HiveTap: strict mode: comparing existing hive table with table descriptor
15/07/08 18:34:26 WARN hadoop.HadoopStepStats: fetching counters timed out after: 5 seconds, attempts: 1
15/07/08 18:34:27 INFO Configuration.deprecation: io.bytes.per.checksum is deprecated. Instead, use dfs.bytes-per-checksum
15/07/08 18:34:27 INFO util.HadoopUtil: using default application jar, may cause class not found exceptions on the cluster
15/07/08 18:34:27 INFO planner.HadoopPlanner: using application jar: /disk1/bigdata/dev/scripts/farlo/1.0.0-20150624014544-cd4bd39/etl-farlo-clm-assembly-1.0.0-20150708030821-b6ae56e.jar
15/07/08 18:34:27 INFO flow.Flow: [] starting
15/07/08 18:34:27 INFO flow.Flow: []  source: Hfs["ParquetScroogeScheme[[UNKNOWN]->[ALL]]"]["/dev/view/intermediate/farlo/clm/ZBP_ID_history"]
15/07/08 18:34:27 INFO flow.Flow: []  source: Hfs["ParquetScroogeScheme[[UNKNOWN]->[ALL]]"]["/dev/view/intermediate/farlo/clm/ZBP_ID_history"]
15/07/08 18:34:27 INFO flow.Flow: []  source: Hfs["CombinedSequenceFileScheme[[0:1]]"]["/tmp/hadoop-acoe_omnds_dintg/maestro/c7e86177-a3a0-4363-8695-5cb4172d6861.seq"]
15/07/08 18:34:27 INFO flow.Flow: []  sink: Hfs["ParquetScroogeScheme[[UNKNOWN]->[0]]"]["/dev/etl/outbound/farlo/clm/v_CLNT_GRUP_STGY_HIST/year=2015/month=07/day=06"]
15/07/08 18:34:27 INFO flow.Flow: []  parallel execution is enabled: true
15/07/08 18:34:27 INFO flow.Flow: []  starting jobs: 1
15/07/08 18:34:27 INFO flow.Flow: []  allocating threads: 1
15/07/08 18:34:27 INFO flow.FlowStep: [] starting step: (1/1) ...year=2015/month=07/day=06
15/07/08 18:34:38 INFO mapred.FileInputFormat: Total input paths to process : 108
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
15/07/08 18:34:42 INFO mapred.FileInputFormat: Total input paths to process : 3
15/07/08 18:34:54 INFO flow.FlowStep: [] submitted hadoop job: job_201506272351_50211
15/07/08 18:34:54 INFO flow.FlowStep: [] tracking url: http://r01hn02.nw-omnia.cba:50030/jobdetails.jsp?jobid=job_201506272351_50211
15/07/08 18:37:31 WARN hadoop.HadoopStepStats: fetching counters timed out after: 5 seconds, attempts: 1
15/07/08 18:44:07 WARN hadoop.HadoopStepStats: fetching counters timed out after: 5 seconds, attempts: 2
15/07/08 18:46:39 WARN hadoop.HadoopStepStats: fetching counters timed out after: 5 seconds, final attempt: 3
```

**Comment by rakheshrnair** [comment]
@maxmithun I am also getting the same exception and believe its the environment issue. Sometimes after rerunning, the jobs are getting succeeded. Also the parameter set to 60 is available from `1.13.0-20150604153322-aff39de` of etl-util

**Comment by vineethvarghese** [comment]
The fix to increase the wait time is in maestro. Are you guys saying that
this fix is not really increasing the timeout? Or is that you haven't
upgraded to the latest maestro version? Would be great if    you guys  can
update this issue with that information too.

Unfortunately the platform team hasn't got back with any information as to
why this happens on our cluster.

**Comment by vineethvarghese** [comment]
@maxmithun You don't have the version of maestro that has this fix. I just verified that etl-util 0.13.0.*** comes with the fix.

**Comment by maxmithun** [comment]
@vineethvarghese  I am closing the issue, since with the new version `1.14.1-20150703071108-fb2a434` we couldnt replicate it.

**Comment by vineethvarghese** [comment]
Thank you
